staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -225,17 +228,61 @@ func (wc *watchChan) RequestWatchProgress() error {
 func (wc *watchChan) sync() error {
 	opts := []clientv3.OpOption{}
 	if wc.recursive {
-		opts = append(opts, clientv3.WithPrefix())
+		if wc.watcher.pagingEnabled {
+			opts = append(opts, clientv3.WithLimit(maxLimit))
+			rangeEnd := clientv3.GetPrefixRangeEnd(wc.key)
+			opts = append(opts, clientv3.WithRange(rangeEnd))
+		} else {
+			opts = append(opts, clientv3.WithPrefix())
+		}
 	}
-	getResp, err := wc.watcher.client.Get(wc.ctx, wc.key, opts...)
-	if err != nil {
-		return err
+
+	var err error
+	var lastKey []byte
+	var getResp *clientv3.GetResponse
+	var withRev int64
+
+	metricsOp := "get"
+	if wc.recursive {
+		metricsOp = "list"
 	}
-	wc.initialRev = getResp.Header.Revision
-	for _, kv := range getResp.Kvs {
-		wc.sendEvent(parseKV(kv))
+
+	preparedKey := wc.key
+
+	// loop until there are no more results
+	for {
+		startTime := time.Now()
+		getResp, err = wc.watcher.client.KV.Get(wc.ctx, preparedKey, opts...)
+		metrics.RecordEtcdRequest(metricsOp, wc.watcher.groupResource.String(), err, startTime)
+		if err != nil {
+			return err
+		}
+
+		if len(getResp.Kvs) == 0 && getResp.More {
+			return fmt.Errorf("no results were found, but etcd indicated there were more values remaining")
+		}
+
+		// send items from the response until no more results
+		for i, kv := range getResp.Kvs {
+			lastKey = kv.Key
+			wc.sendEvent(parseKV(kv))
+			// free kv early. Long lists can take O(seconds) to decode.
+			getResp.Kvs[i] = nil
+		}
+
+		wc.initialRev = getResp.Header.Revision
+
+		// no more results remain or we didn't request paging
+		if !getResp.More || !wc.watcher.pagingEnabled {
+			return nil
+		}
+
+		preparedKey = string(lastKey) + "\x00"
+		if withRev == 0 {
+			withRev = getResp.Header.Revision
+			opts = append(opts, clientv3.WithRev(withRev))
+		}
```

Comment:
Please add a unit test that exercises the new logic.
The two main things that I would like to ensure are validated in this test are:
- the fact that we page
- the consistency [i.e. add a new item to etcd before returning a second page and ensure that it's not returned]

Reply:
Could you give some advice? We are constantly getting and sending events in the sync function, which is difficult to guarantee the order of returning the second page of data and adding new objects

Reply:
An an example, a wrapper around etcd client that does that on response on calls.

Reply:
Thanks for your advice, Tests have been added